### PR TITLE
fixing product categories to include more realistic text

### LINKED ISSUE #54: Use more realistic product categories
It looks like smooth generator uses a list of real "products" when generating - e.g. hat, glove, knife etc.

The categories look like they are more randomly generated lorem-ipsum style text (e.g. `est, nesciunt, numquam, quia, repellendus`.

It would be cool if the the product type was used for categories also, e.g. all hats in a `Hats` category, for example:

Screen Shot 2021-04-12 at 9 20 08 AM

Could also add categories for the other words used in the product name, e.g. the second "material" word - `Leather`, `Concrete`, `Wooden` etc.

Using consistent categories like this would make it easier to test different scenarios involving product categories, as the expected category would be more intuitive.

### PR DESCRIPTION
**Pull request by ovidiul**
Closes #54 

This PR updates product category name and product tag generation to use more realistic names.

For product categories, the terms are extracted from the generated product name except for the first occurrence, while for product tags, the data is appended to the existing generated tags.

Screenshot 2021-09-22 at 10 26 15

The [faker-provide-collection](https://github.com/mbezhanov/faker-provider-collection) also provides a `$faker->department` method to get a more realistic category name, so that is also used to extend the categories and tags list.

### Changelog entry

> Add: use product name to generate more realistic product term names.

### CODE REVIEW

**Comment by rrennick** [comment]
> Closes #54 

@ovidiul Thanks for the PR. I've added it to our review queue. I edited the description and included the above line. With this line Github will automatically close the issue when the PR is merged.

**Comment by rrennick** [comment]
@ovidiul Some comments for discussion after generating 500 products:

- This looks like it duplicates tags and categories:
```
$ wp wc product_tag list --orderby=name --order=desc --user=admin
...
 873  | shoes &amp; toys        | shoes-toys         |             | 1     |
| 752  | shoes &amp; tools       | shoes-tools        |             | 1     |
| 704  | shoes &amp; sports      | shoes-sports       |             | 1     |
| 923  | shoes &amp; jewelry     | shoes-jewelry      |             | 1     |
| 663  | shoes &amp; industrial  | shoes-industrial   |             | 2     |
| 1109 | shoes &amp; grocery     | shoes-grocery      |             | 0     |
| 902  | shoes &amp; electronics | shoes-electronics  |             | 1     |
| 823  | shoes &amp; automotive  | shoes-automotive   |             | 1     |
| 599  | shoes                   | shoes              |             | 31    |
...
$ wp wc product_cat list --orderby=name --order=desc --user=admin
...
| 1111 | Shoes &amp; Sports                          | shoes-sports                                 | 0      |             | default |       | 0          | 1     |
| 919  | Shoes &amp; Kids                            | shoes-kids                                   | 0      |             | default |       | 0          | 1     |
| 946  | Shoes &amp; Jewelry                         | shoes-jewelry                                | 0      |             | default |       | 0          | 0     |
| 1043 | Shoes &amp; Industrial                      | shoes-industrial                             | 0      |             | default |       | 0          | 1     |
| 746  | Shoes &amp; Home                            | shoes-home                                   | 0      |             | default |       | 0          | 1     |
| 768  | Shoes &amp; Garden                          | shoes-garden                                 | 0      |             | default |       | 0          | 1     |
| 585  | Shoes &amp; Games                           | shoes-games                                  | 0      |             | default |       | 0          | 1     |
| 1021 | Shoes &amp; Electronics                     | shoes-electronics                            | 0      |             | default |       | 0          | 1     |
| 962  | Shoes &amp; Computers                       | shoes-computers                              | 0      |             | default |       | 0          | 1     |
| 1097 | Shoes &amp; Books                           | shoes-books                                  | 0      |             | default |       | 0          | 0     |
| 597  | Shoes                                       | shoes                                        | 0      |             | default |       | 0          | 26    |
...
```

Do you think it would work better if they were split in some way (eg. first word in product name is used for category, remainder for tags)?

- For tags in particular, I would expect `shoes, sports` vs `shoes, shoes & sports`. But this could also be applied to categories where in a live site the categories would likely be hierarchical `Shoes, Shoes > Sports`.

Outside of those two things the code looks great. I'm going to edit the PR description to include the changelog entry.

**Comment by ovidiul** [comment]
@rrennick it makes sense to me to extract the words only from a "word1 & word2" expression.

However, taking this product name "Lightweight Wooden Table", it makes more sense to me to put the first word into a tag(as that might indicate to be a particularity of the product) and the rest into categories(with a hierarchy included), but we can go either way.

My latest commit adds these changes including the hierarchy for categories and `str_word_count` as a way to better extract words from the generated names.


Screenshot 2021-09-29 at 11 48 11
Screenshot 2021-09-29 at 11 48 25

Thanks!

File: includes/Generator/Generator.php [hunk 1]
```diff
@@ -68,30 +69,57 @@ protected static function generate_term_ids( $limit, $taxonomy ) {
 			return $term_ids;
 		}
 
-		$terms = self::$faker->words( $limit );
+		$words       = str_word_count( $name, 1 );
+		$extra_terms = str_word_count( implode( ' ', self::$faker->department( $limit ) ), 1 );
```
Comment (rrennick) [inline includes/Generator/Generator.php]: @ovidiul `self::$faker->department( $limit )` returns a string vs an array causing the `implode` to throw a warning.

Comment (ovidiul) [reply includes/Generator/Generator.php]: fixed, my bad!

**Comment by rrennick** [comment]
@ovidiul The categories look good but there seems to be an issue with the tags:

```
$ wp wc product list --orderby=id --order=desc --fields=id,name,categories,tags --user=admin
+-------+-----------------------------+------------------------------------------------------------+--------------------------------------------------------------+
| id    | name                        | categories                                                 | tags                                                         |
+-------+-----------------------------+------------------------------------------------------------+--------------------------------------------------------------+
| 18722 | Lightweight Linen Bench     | [{"id":1149,"name":"Bench","slug":"bench-linen"},{"id":610 | [{"id":723,"name":"a","slug":"a"},{"id":237,"name":"error"," |
|       |                             | ,"name":"Linen","slug":"linen"},{"id":2490,"name":"Shoes", | slug":"error"},{"id":269,"name":"illo","slug":"illo"},{"id": |
|       |                             | "slug":"shoes-bench-linen"}]                               | 402,"name":"iure","slug":"iure"},{"id":496,"name":"lightweig |
|       |                             |                                                            | ht","slug":"lightweight"},{"id":201,"name":"non","slug":"non |
|       |                             |                                                            | "}]                                                          |
| 18721 | Enormous Cotton Bench       | [{"id":2487,"name":"Beauty","slug":"beauty-bench-cotton"}, | [{"id":351,"name":"ab","slug":"ab"},{"id":503,"name":"enormo |
|       |                             | {"id":2090,"name":"Bench","slug":"bench-cotton"},{"id":672 | us","slug":"enormous"},{"id":322,"name":"perspiciatis","slug |
|       |                             | ,"name":"Cotton","slug":"cotton"},{"id":2489,"name":"Healt | ":"perspiciatis"},{"id":357,"name":"placeat","slug":"placeat |
|       |                             | h","slug":"health-outdoors-beauty-bench-cotton"},{"id":248 | "}]                                                          |
|       |                             | 8,"name":"Outdoors","slug":"outdoors-beauty-bench-cotton"} |                                                              |
```

**Comment by ovidiul** [comment]
@rrennick Can you help me understand the issues with the tags? 

As before, the tags are generated from the `lipsum ipsurem` text are now appended the first word of each product.

I am not sure if we should add similar text to the categories here?

**Comment by rrennick** [comment]
> As before, the tags are generated from the lipsum ipsurem text are now appended the first word of each product.

I wasn't expecting `error` would be generated by faker. But, that could have been generated in other testing. I tried this on a fresh install and the tags look okay.

In testing in the fresh install I noticed that this branch was generating an average of 2-3 unique categories per product with lots of duplicate categories:

Screen Shot 2021-10-04 at 12 39 19 PM

I removed the hierarchical logic with the diff below and ran a couple tests. With these changes categories were reused and applied to multiple products.

```
$ git diff
diff --git a/includes/Generator/Generator.php b/includes/Generator/Generator.php
index 6414dcd..5b79c24 100644
--- a/includes/Generator/Generator.php
+++ b/includes/Generator/Generator.php
@@ -79,17 +79,10 @@ abstract class Generator {
                        $terms = array_merge( self::$faker->words( $limit ), array( strtolower( $words[0] ) ) );
                }
 
-               $args = array( 'parent' => 0 );
-
                foreach ( $terms as $term ) {
-                       if ( isset( self::$term_ids[ $taxonomy ][ $args['parent'] ], self::$term_ids[ $taxonomy ][ $args['parent'] ][ $term ] ) ) {
-                               $term_id    = self::$term_ids[ $taxonomy ][ $args['parent'] ][ $term ];
+                       if ( isset( self::$term_ids[ $taxonomy ], self::$term_ids[ $taxonomy ][ $term ] ) ) {
+                               $term_id    = self::$term_ids[ $taxonomy ][ $term ];
                                $term_ids[] = $term_id;
-
-                               if ( is_taxonomy_hierarchical( $taxonomy ) ) {
-                                       $args['parent'] = $term_id;
-                               }
-
                                continue;
                        }
 
@@ -97,7 +90,6 @@ abstract class Generator {
                        $args    = array(
                                'taxonomy' => $taxonomy,
                                'name'     => $term,
-                               'parent'   => $args['parent'],
                        );
 
                        $existing = get_terms( $args );
@@ -115,11 +107,7 @@ abstract class Generator {
                        if ( $term_id ) {
 
                                $term_ids[] = $term_id;
-                               self::$term_ids[ $taxonomy ][ $args['parent'] ][ $term ] = $term_id;
-
-                               if ( is_taxonomy_hierarchical( $taxonomy ) ) {
-                                       $args['parent'] = $term_id;
-                               }
+                               self::$term_ids[ $taxonomy ][ $term ] = $term_id;
                        }
                }
```

> But this could also be applied to categories where in a live site the categories would likely be hierarchical Shoes, Shoes > Sports.

I apologize. This comment was a bit misleading. I was hoping to point out that a realistic shop would be unlikely to have nearly duplicating categories vs suggesting that we needed to incorporate category hierarchy.

```
 873  | shoes &amp; toys        | shoes-toys         |             | 1     |
| 752  | shoes &amp; tools       | shoes-tools        |             | 1     |
| 704  | shoes &amp; sports      | shoes-sports       |             | 1     |
| 923  | shoes &amp; jewelry     | shoes-jewelry      |             | 1     |
| 663  | shoes &amp; industrial  | shoes-industrial   |             | 2     |
| 1109 | shoes &amp; grocery     | shoes-grocery      |             | 0     |
| 902  | shoes &amp; electronics | shoes-electronics  |             | 1     |
| 823  | shoes &amp; automotive  | shoes-automotive   |             | 1     |
| 599  | shoes                   | shoes              |             | 31    |
```

**Comment by ovidiul** [comment]
@rrennick 👋 

I understand now, I've updated the PR to remove the hierarchies duplication.

One small change to your diff is adding this line  https://github.com/woocommerce/wc-smooth-generator/pull/65/files#diff-89881f441fadd760340e944c92565542fed45b5412ef62b197decabea0ead89cR85 to provide better consistency for categories.

On the `error` tag, I was not able to duplicate it either, but curious how that ended that initially 😄.

Thanks for reviewing this, I really appreciate it.

**Review by rrennick — APPROVED** [review]
@ovidiul Nice work on this. Thanks for the help.

**Pull request merged**